Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -411,155 +411,152 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
 
     RefPtr decompressionSession = this->decompressionSession();
 
-    if (m_isDecodingSample || m_gotDecodingError || !decompressionSession)
+    if (m_gotDecodingError || !decompressionSession)
         return;
 
-    if (m_compressedSampleQueue.isEmpty())
-        return;
-
-    RetainPtr timebase = this->timebase();
-    MediaTime currentTime = timebase ? PAL::toMediaTime(PAL::CMTimebaseGetTime(timebase.get())) : MediaTime::invalidTime();
-    double playbackRate = timebase ? PAL::CMTimebaseGetRate(timebase.get()) : 0;
-
-    WebCoreDecompressionSession::DecodingFlags decodingFlags;
-
-    if (currentTime.isValid() && !m_wasProtected && !m_decompressionSessionWasBlocked) {
-        auto lowWaterMarkTime = currentTime + DecodeLowWaterMark;
-        auto highWaterMarkTime = currentTime + DecodeHighWaterMark;
-        auto endTime = lastDecodedSampleTime();
-        if (endTime.isValid() && endTime > highWaterMarkTime) {
-            auto [sample, upcomingMinimum, flushId, blocked] = m_compressedSampleQueue.first();
-            upcomingMinimum = std::min(sample->presentationTime(), upcomingMinimum.isValid() ? upcomingMinimum : MediaTime::positiveInfiniteTime());
-
-            if (endTime < upcomingMinimum) {
-                if (m_lastMinimumUpcomingPresentationTime.isInvalid() || upcomingMinimum != m_lastMinimumUpcomingPresentationTime) {
-                    ASSERT(m_lastMinimumUpcomingPresentationTime.isInvalid() || m_lastMinimumUpcomingPresentationTime < upcomingMinimum);
-                    m_lastMinimumUpcomingPresentationTime = upcomingMinimum;
-                    LogPerformance("VideoMediaSampleRenderer::decodeNextSampleIfNeeded currentTime:%0.2f expectMinimumUpcomingSampleBufferPresentationTime:%0.2f decoded queued:%zu upcoming:%zu high watermark reached", currentTime.toDouble(), m_lastMinimumUpcomingPresentationTime.toDouble(), decodedSamplesCount(), m_compressedSampleQueue.size());
-                    [rendererOrDisplayLayer() expectMinimumUpcomingSampleBufferPresentationTime:PAL::toCMTime(m_lastMinimumUpcomingPresentationTime)];
+    while (!m_compressedSampleQueue.isEmpty() && !m_isDecodingSample) {
```

Comment:
Can code inside the loop modify (add to) `m_compressedSampleQueue`?

If so, that would have lead to infinite recursion previously.  (I think this is better as an iterative loop than using recursion, though.)